[Style] Convert some OptionSet based properties to strong style types

weinig opened the pull request:
#### 56d8edca88cb6166aac1459c1e1d01a0a51a32d5
<pre>
[Style] Convert some OptionSet based properties to strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=296333">https://bugs.webkit.org/show_bug.cgi?id=296333</a>

Reviewed by NOBODY (OOPS!).

Adds a new aggregate, `CSS::PrimitiveKeywordSet`, that acts just like
an OptionSet, but instead of using an `enum`, it uses a set of CSS::Keyword
constants. This allows for completely generating the conversion and
serialization in the general case.

Adopts new type for `text-decoration-line`, `position-visibility`,
`margin-trim`, and `-webkit-line-box-contain`.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
* Source/WebCore/accessibility/AXObjectCache.cpp:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/scripts/process-css-values.py:
* Source/WebCore/css/values/primitives/CSSPrimitiveKeywordSet.h: Added.
* Source/WebCore/editing/Editor.cpp:
* Source/WebCore/editing/cocoa/EditingHTMLConverter.mm:
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBoxInlines.h:
* Source/WebCore/layout/formattingContexts/inline/InlineLineBoxBuilder.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineQuirks.cpp:
* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayContentBuilder.cpp:
* Source/WebCore/rendering/RenderBlock.cpp:
* Source/WebCore/rendering/RenderBlock.h:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
* Source/WebCore/rendering/RenderBlockFlow.h:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderBoxInlines.h:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
* Source/WebCore/rendering/RenderFlexibleBox.h:
* Source/WebCore/rendering/RenderGrid.cpp:
* Source/WebCore/rendering/RenderGrid.h:
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/StyledMarkedText.cpp:
* Source/WebCore/rendering/TextBoxPainter.cpp:
* Source/WebCore/rendering/TextDecorationPainter.cpp:
* Source/WebCore/rendering/TextDecorationPainter.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleDifference.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/svg/SVGTextBoxPainter.cpp:
* Source/WebCore/rendering/svg/SVGTextBoxPainter.h:
* Source/WebCore/rendering/updating/RenderTreeBuilderFirstLetter.cpp:
* Source/WebCore/style/InlineTextBoxStyle.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
* Source/WebCore/style/values/anchor-position/StylePositionVisibility.h: Added.
* Source/WebCore/style/values/box/StyleMarginTrim.cpp: Added.
* Source/WebCore/style/values/box/StyleMarginTrim.h: Added.
* Source/WebCore/style/values/inline/StyleLineBoxContain.h:
* Source/WebCore/style/values/primitives/StylePrimitiveKeyword+CSSValueConversion.h:
* Source/WebCore/style/values/primitives/StylePrimitiveKeywordSet+CSSValueConversion.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveKeywordSet+CSSValueCreation.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveKeywordSet+Logging.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveKeywordSet+Serialization.h: Added.
* Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.h: Added.
* Source/WebKitLegacy/mac/WebView/WebView.mm:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/56d8edca88cb6166aac1459c1e1d01a0a51a32d5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112862 "6 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32597 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23075 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119066 "Hash 56d8edca for PR 48378 does not build (failure)") | [❌ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63389 "Hash 56d8edca for PR 48378 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114824 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33249 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41160 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/119066 "Hash 56d8edca for PR 48378 does not build (failure)") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/63389 "Hash 56d8edca for PR 48378 does not build (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115809 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26542 "Found 60 new test failures: accessibility/checkbox-radio-element-rect.html accessibility/visible-character-range-vertical-rl.html compositing/backing/whitespace-nodes-no-backing.html compositing/direct-image-compositing.html compositing/generated-content.html compositing/geometry/root-layer-update.html compositing/geometry/transfrom-origin-on-zero-size-layer.html compositing/sibling-positioning.html css1/basic/containment.html css1/basic/contextual_selectors.html ... (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101531 "Found 1 new API test failure: TestWebKitAPI.WKWebView.RequestAllTextRunsWithSubframes (failure)") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/119066 "Hash 56d8edca for PR 48378 does not build (failure)") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25830 "Found 60 new test failures: imported/w3c/web-platform-tests/compat/webkit-text-fill-color-property-001c.html imported/w3c/web-platform-tests/css/CSS2/floats/floats-placement-vertical-001b.xht imported/w3c/web-platform-tests/css/CSS2/floats/floats-placement-vertical-001c.xht imported/w3c/web-platform-tests/css/CSS2/tables/caption-side-applies-to-001.xht imported/w3c/web-platform-tests/css/CSS2/tables/caption-side-applies-to-005.xht imported/w3c/web-platform-tests/css/CSS2/tables/collapsing-border-model-010a.xht imported/w3c/web-platform-tests/css/CSS2/tables/collapsing-border-model-010b.xht imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-vlr.html imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-vrl.html imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-computed.html ... (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19663 "Found 60 new test failures: css1/conformance/forward_compatible_parsing.html css1/font_properties/font.html css1/formatting_model/inline_elements.html css1/text_properties/line_height.html css2.1/20110323/absolute-non-replaced-height-002.htm css2.1/20110323/absolute-non-replaced-height-007.htm css2.1/20110323/absolute-non-replaced-height-009.htm css2.1/20110323/absolute-non-replaced-width-001.htm css2.1/20110323/absolute-non-replaced-width-002.htm css2.1/20110323/absolute-non-replaced-width-003.htm ... (failure)") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62824 "Hash 56d8edca for PR 48378 does not build (failure)") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95936 "Found 11 new API test failures: TestWebKitAPI.EditorStateTests.MarkedTextRange_VerticalRangeSelection, TestWebKitAPI.AccessibilityTests.RectsForSpeakingSelectionBasic, TestWebKitAPI.EditorStateTests.MarkedTextRange_VerticalCaretSelection, TestWebKitAPI.AccessibilityTests.StoreSelection, TestWebKitAPI.RequestTextInputContext.FocusedEditableLineWithOnlyLineBreak, TestWebKitAPI.AccessibilityTests.RectsForSpeakingSelectionWithLineWrapping, TestWebKitAPI.DragAndDropTests.LinkToInput, TestWebKitAPI.FontAttributes.FontAttributesAfterChangingSelection, TestWebKitAPI.WKWebViewEditActions.ToggleStrikeThrough, TestWebKitAPI.EditorStateTests.TypingAttributesUnderline ... (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19737 "Found 60 new test failures: css1/conformance/forward_compatible_parsing.html css1/font_properties/font.html css1/formatting_model/inline_elements.html css1/text_properties/line_height.html css2.1/20110323/absolute-non-replaced-height-002.htm css2.1/20110323/absolute-non-replaced-height-007.htm css2.1/20110323/absolute-non-replaced-height-009.htm css2.1/20110323/absolute-non-replaced-width-001.htm css2.1/20110323/absolute-non-replaced-width-002.htm css2.1/20110323/absolute-non-replaced-width-003.htm ... (failure)") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122287 "Hash 56d8edca for PR 48378 does not build (failure)") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39940 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29784 "Found 11 new test failures: editing/execCommand/toggle-mixed-text-decorations.html fast/block/lineboxcontain/font-replaced.html fast/css3-text/css3-text-decoration/getComputedStyle/getComputedStyle-text-decoration-line.html fast/css3-text/css3-text-decoration/text-decoration-line.html imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-computed.html imported/w3c/web-platform-tests/css/css-box/inheritance.html imported/w3c/web-platform-tests/css/css-box/parsing/margin-trim-computed.html imported/w3c/web-platform-tests/css/css-text-decor/parsing/text-decoration-line-computed.html imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-line.html imported/w3c/web-platform-tests/css/css-text-decor/text-underline-position-horizontal.html ... (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/122287 "Hash 56d8edca for PR 48378 does not build (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40324 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97751 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/122287 "Hash 56d8edca for PR 48378 does not build (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39627 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17439 "Found 60 new test failures: compositing/scrolling/async-overflow-scrolling/negative-z-in-scroller.html css1/conformance/forward_compatible_parsing.html css1/font_properties/font.html css1/formatting_model/inline_elements.html css1/text_properties/line_height.html css2.1/20110323/absolute-non-replaced-height-002.htm css2.1/20110323/absolute-non-replaced-height-007.htm css2.1/20110323/absolute-non-replaced-height-009.htm css2.1/20110323/absolute-non-replaced-width-001.htm css2.1/20110323/absolute-non-replaced-width-002.htm ... (failure)") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36051 "Hash 56d8edca for PR 48378 does not build (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39826 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45324 "Found 2 new failures in style/values/primitives/StylePrimitiveKeywordSet+CSSValueCreation.h, style/values/primitives/StylePrimitiveKeywordSet+Serialization.h and found 2 fixed files: style/values/primitives/StylePrimitiveKeywordSet+CSSValueCreation.h, style/values/primitives/StylePrimitiveKeywordSet+Serialization.h") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39466 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42799 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41204 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1f9d60fdabf6359ea8608eeb04dae0ef6093b27e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48378

weinig committed:
[Style] Convert some OptionSet based properties to strong style types
https://bugs.webkit.org/show_bug.cgi?id=296333

Reviewed by NOBODY (OOPS!).

Adds a new aggregate, `CSS::PrimitiveKeywordSet`, that acts just like
an OptionSet, but instead of using an `enum`, it uses a set of CSS::Keyword
constants. This allows for completely generating the conversion and
serialization in the general case.

Adopts new type for `text-decoration-line`, `position-visibility`,
`margin-trim`, and `-webkit-line-box-contain`.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
* Source/WebCore/accessibility/AXObjectCache.cpp:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/scripts/process-css-values.py:
* Source/WebCore/css/values/primitives/CSSPrimitiveKeywordSet.h: Added.
* Source/WebCore/editing/Editor.cpp:
* Source/WebCore/editing/cocoa/EditingHTMLConverter.mm:
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBoxInlines.h:
* Source/WebCore/layout/formattingContexts/inline/InlineLineBoxBuilder.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineQuirks.cpp:
* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayContentBuilder.cpp:
* Source/WebCore/rendering/RenderBlock.cpp:
* Source/WebCore/rendering/RenderBlock.h:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
* Source/WebCore/rendering/RenderBlockFlow.h:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderBoxInlines.h:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
* Source/WebCore/rendering/RenderFlexibleBox.h:
* Source/WebCore/rendering/RenderGrid.cpp:
* Source/WebCore/rendering/RenderGrid.h:
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/StyledMarkedText.cpp:
* Source/WebCore/rendering/TextBoxPainter.cpp:
* Source/WebCore/rendering/TextDecorationPainter.cpp:
* Source/WebCore/rendering/TextDecorationPainter.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleDifference.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/svg/SVGTextBoxPainter.cpp:
* Source/WebCore/rendering/svg/SVGTextBoxPainter.h:
* Source/WebCore/rendering/updating/RenderTreeBuilderFirstLetter.cpp:
* Source/WebCore/style/InlineTextBoxStyle.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
* Source/WebCore/style/values/anchor-position/StylePositionVisibility.h: Added.
* Source/WebCore/style/values/box/StyleMarginTrim.cpp: Added.
* Source/WebCore/style/values/box/StyleMarginTrim.h: Added.
* Source/WebCore/style/values/inline/StyleLineBoxContain.h:
* Source/WebCore/style/values/primitives/StylePrimitiveKeyword+CSSValueConversion.h:
* Source/WebCore/style/values/primitives/StylePrimitiveKeywordSet+CSSValueConversion.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveKeywordSet+CSSValueCreation.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveKeywordSet+Logging.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveKeywordSet+Serialization.h: Added.
* Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.h: Added.
* Source/WebKitLegacy/mac/WebView/WebView.mm:

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e1100e173d0e88d547d0ef092c9511a0e77daafb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48378

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/56d8edca88cb6166aac1459c1e1d01a0a51a32d5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48378
